### PR TITLE
ECER-1223: Opt-out references/ submit character reference BE piece

### PR DESCRIPTION
**Pull request by farzadnadiri**
---
name: Opt-out references/ submit character reference
about: opt-out functionality implemented for both character and work experience references
---

## Title
ECER-1223: Opt-out references/ submit character reference

## Description

- Opt-out implemented for character reference
- Opt-out implemented for work experience reference
- Submit character reference backend piece

## Related Jira Issue(s)

- ECER-1223
- ECER-1226

## Checklist
- [x] I have tested these changes locally.
- [x] Changes to backend endpoints are covered by passing integration tests.
- [ ] I have added or updated the necessary documentation.

### CODE REVIEW

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/ConfigurationEndpoints.cs [hunk 1]
```diff
@@ -13,6 +16,14 @@ public void Register(IEndpointRouteBuilder endpointRouteBuilder)
       var appConfig = configuration.Get<ApplicationConfiguration>();
       return TypedResults.Ok(appConfig);
     }).WithOpenApi("Returns the UI initial configuration", string.Empty, "configuration_get");
+
+    endpointRouteBuilder.MapGet("/api/provincelist", async (HttpContext ctx, IMediator messageBus, IMapper mapper, CancellationToken ct) =>
+    {
+      var results = await messageBus.Send(new ProvincesQuery(), ct);
+      return TypedResults.Ok(mapper.Map<IEnumerable<Province>>(results.Items));
+    }).WithOpenApi("Handles province queries", string.Empty, "province_get")
+      .RequireAuthorization()
```
Comment (ytqsl) [inline src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/ConfigurationEndpoints.cs:25]: Not sure if you want to protect this endpoint. Also, it is worth enabling output caching and/or cache the list in the distributed cache.

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/ConfigurationEndpoints.cs [hunk 2]
```diff
@@ -32,3 +43,5 @@ public record OidcAuthenticationSettings
   public string Scope { get; set; } = null!;
   public string? Idp { get; set; }
 }
+
+public record Province(string ProvinceId, string ProvinceName);
```
Comment (ytqsl) [inline src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/ConfigurationEndpoints.cs:47]: Do you want the front end to know about the ID or is it just the standard abbreviation?

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/References/ReferencesEndpoints.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      if (request.Token == null) return TypedResults.BadRequest("Token is required");
+      var result = await messageBus.Send(mapper.Map<Managers.Registry.Contract.Applications.CharacterReferenceSubmissionRequest>(request), ct);
+      if (!result.IsSuccess)
+      {
+        return TypedResults.BadRequest(result.ErrorMessage);
+      }
+      return TypedResults.Ok();
+    }).WithOpenApi("Handles character reference submission", string.Empty, "character_reference_post").WithParameterValidation();
+
+    endpointRouteBuilder.MapPost("/api/OptOutReference", async Task<Results<Ok, BadRequest<string>>> (OptOutReferenceRequest request, IMediator messageBus, HttpContext httpContext, IMapper mapper, CancellationToken ct) =>
```
Comment (ytqsl) [inline src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/References/ReferencesEndpoints.cs:23]: I'm not convinced this is how you'd want to structure the API, it is very RPC style right now. I also assume we have backlog stories to add recaptcha protections to these endpoint, right?

File: src/ECER.Managers.Registry.Contract/PortalInvitations/Contract.cs [hunk 4]
```diff
@@ -21,13 +22,25 @@ public record PortalInvitation(string? Id, string Name, string ReferenceFirstNam
   public string? ApplicantFirstName { get; set; }
   public string? ApplicantLastName { get; set; }
   public string? ApplicationId { get; set; }
+  public IEnumerable<CertificationType>? CertificationTypes { get; set; }
```
Comment (ytqsl) [inline src/ECER.Managers.Registry.Contract/PortalInvitations/Contract.cs:25]: It is safe to set the default value for enumerables to array.empty, it reduces the amount of null checks you need to do later

File: src/ECER.Managers.Registry/ApplicationHandlers.cs [hunk 5]
```diff
@@ -128,4 +140,84 @@ public async Task<ApplicationsQueryResults> Handle(ApplicationsQuery request, Ca
...
+  /// <summary>
+  /// Handles Reference Opt out
+  /// </summary>
+  /// <param name="request"></param>
+  /// <param name="cancellationToken"></param>
+  /// <returns></returns>
+  /// <exception cref="InvalidCastException"></exception>
+  public async Task<ReferenceSubmissionResult> Handle(Contract.Applications.OptOutReferenceRequest request, CancellationToken cancellationToken)
+  {
+    ArgumentNullException.ThrowIfNull(portalInvitationRepository);
```
Comment (ytqsl) [inline src/ECER.Managers.Registry/ApplicationHandlers.cs:193]: btw all these checks can be deleted now as they're injected in the ctor

File: src/ECER.Resources.Documents/Configurer.cs [hunk 6]
```diff
@@ -12,5 +13,6 @@ public void Configure([NotNull] ConfigurationContext configurationContext)
   {
     configurationContext.Services.AddTransient<IApplicationRepository, ApplicationRepository>();
     configurationContext.Services.AddTransient<IPortalInvitationRepository, PortalInvitationRepository>();
+    configurationContext.Services.AddTransient<IMetadataResourceRepository, MetadataResourceRepository>();
```
Comment (ytqsl) [inline src/ECER.Resources.Documents/Configurer.cs:16]: The metadata should be moved to its own project called metadata resource

**Review by ytqsl — COMMENTED** [review]
Some comments, most can be worked on later as refactoring of tech debt

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/ConfigurationEndpoints.cs (hunk 2 above)
Comment (farzadnadiri) [reply src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/ConfigurationEndpoints.cs:47]: It is the id from dynamics like all other lookup values

File: src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/References/ReferencesEndpoints.cs (hunk 3 above)
Comment (farzadnadiri) [reply src/ECER.Clients.RegistryPortal/ECER.Clients.RegistryPortal.Server/References/ReferencesEndpoints.cs:23]: Ya, We have stories to protect these endpoints using a captcha

**Review by SoLetsDev — APPROVED** [review]

**Pull request merged**